subsys/bluetooth/controller/ll_sw/ull_df.c

```diff
@@ -666,3 +601,204 @@ static struct lll_df_adv_cfg *df_adv_cfg_acquire(void)
+uint8_t ll_df_set_conn_cte_req_enable(uint16_t handle, uint8_t enable, uint8_t cte_request_interval,
+				      uint8_t requested_cte_length, uint8_t requested_cte_type)
+{
+	struct ll_conn *conn;
+
+	conn = ll_connected_get(handle);
+	if (!conn) {
+		return BT_HCI_ERR_UNKNOWN_CMD;
+	}
+
+	if (!enable) {
+		/* There is no parameter validation for disable operation. */
+
+		/* TODO: Add missing implementation of disable CTE reques.
+		 * Requires refactored LLCPs.
+		 */
+	} else {
+		if (conn->df_rx_params.state == DF_CTE_SAMPLING_UNINITIALIZED) {
+			return BT_HCI_ERR_CMD_DISALLOWED;
+		}
+
+		/* TODO: check if CTE_REQ LLCP is active. Add when merged with refactored LLCPs */
+
+#if defined(CONFIG_BT_CTLR_PHY)
+		/* Phy may be changed to CODED only if PHY update procedure is supproted. In other
+		 * case the mandatory PHY1M is used (that supports CTE).
+		 */
+		if (conn->lll.phy_tx == PHY_CODED) {
+			return BT_HCI_ERR_CMD_DISALLOWED;
+		}
+#endif /* CONFIG_BT_CTLR_PHY */
+
+		if (cte_request_interval != 0 && cte_request_interval < conn->lll.latency) {
+			return BT_HCI_ERR_CMD_DISALLOWED;
+		}
+
+		if (requested_cte_length < BT_HCI_LE_CTE_LEN_MIN ||
+		    requested_cte_length > BT_HCI_LE_CTE_LEN_MAX) {
+			return BT_HCI_ERR_UNSUPP_FEATURE_PARAM_VAL;
+		}
+
+		if (requested_cte_type != BT_HCI_LE_AOA_CTE &&
+		    requested_cte_type != BT_HCI_LE_AOD_CTE_1US &&
+		    requested_cte_type != BT_HCI_LE_AOD_CTE_2US) {
+			return BT_HCI_ERR_UNSUPP_FEATURE_PARAM_VAL;
+		}
+
+		/* If controller is aware of features supported by peer device then check
+		 * whether required features are enabled.
+		 */
+		if (conn->common.fex_valid &&
+		    (!(conn->llcp_feature.features_peer & BIT64(BT_LE_FEAT_BIT_CONN_CTE_RESP)) ||
+		     ((requested_cte_type == BT_HCI_LE_AOD_CTE_1US ||
+		       requested_cte_type == BT_HCI_LE_AOD_CTE_2US) &&
+		      !(conn->llcp_feature.features_peer &
+			BIT64(BT_LE_FEAT_BIT_ANT_SWITCH_TX_AOD))))) {
+			return BT_HCI_ERR_UNSUPP_REMOTE_FEATURE;
+		}
+	}
+
+	/* TODO: implement disable of the CTE if PHY is changed to coded */
+
+	return BT_HCI_ERR_CMD_DISALLOWED;
+}
+#endif /* CONFIG_BT_CTLR_DF_CONN_CTE_REQ */
+
+/* @brief Function provides information about Direction Finding
+ *        antennas switching and sampling related settings.
+ *
+ * @param[out]switch_sample_rates       Pointer to store available antennas
+ *                                      switch-sampling configurations.
+ * @param[out]num_ant                   Pointer to store number of available
+ *                                      antennas.
+ * @param[out]max_switch_pattern_len    Pointer to store maximum number of
+ *                                      antennas ids in switch pattern.
+ * @param[out]max_cte_len               Pointer to store maximum length of CTE
+ *                                      in [8us] units.
+ */
+void ll_df_read_ant_inf(uint8_t *switch_sample_rates, uint8_t *num_ant,
+			uint8_t *max_switch_pattern_len, uint8_t *max_cte_len)
```

Comment:
There are four arguments in the function declaration. Most architecture (also ARM(32) ABI) should be able to pass them through registers. That would save instructions for data packing and unpacking from structure. This is a reason why I used 4 pointers instead of pointer to structure.